File: app/src/main/java/com/pgssoft/mvvm/services/SchedulerService.java
package com.pgssoft.mvvm.services;

import android.os.Handler;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;

/**
 * Created by bstokrocki on 19.02.2017.
 */

public class SchedulerService {
    private final Handler handler;
    private final HashMap<Runnable, Runnable> tasks;

    public SchedulerService() {
        handler = new Handler();
        tasks = new HashMap<>();
    }

    public void scheduleRepeatingTask(final Runnable task, final long interval) {
        Runnable internalTask = new Runnable() {
            @Override
            public void run() {
                task.run();
                runRepeating(this, interval);
            }
        };

        if (tasks.put(task, internalTask) == null) {
            runRepeating(internalTask, interval);
        }
    }

    public void cancelTask(Runnable task) {
        Runnable toCancel = tasks.get(task);

        if (toCancel != null) {
            handler.removeCallbacks(toCancel);
        }

        tasks.remove(task);
    }

    private void runRepeating(final Runnable task, long interval) {
        handler.postDelayed(task, interval);
    }
}

File: kotlin/app/src/main/java/com/pgssoft/kotlinplayground/service/SchedulerService.kt
package com.pgssoft.kotlinplayground.service

import android.os.Handler
import java.util.*

/**
 * Created by bstokrocki on 19.02.2017.
 */

class SchedulerService {
    private val handler = Handler()
    private val tasks = HashMap<() -> Unit, Runnable>()

    fun scheduleRepeatingTask(task: () -> Unit, interval: Long) {
        val internalTask = object : Runnable {
            override fun run() {
                task.invoke()
                runRepeating(this, interval)
            }
        }

        tasks.put(task, internalTask) ?: runRepeating(internalTask, interval)
    }

    fun cancelTask(task: () -> Unit) {
        val toCancel = tasks[task]

        if (toCancel != null) {
            handler.removeCallbacks(toCancel)
        }

        tasks.remove(task)
    }

    private fun runRepeating(task: Runnable, interval: Long) {
        handler.postDelayed(task, interval)
    }
}
